pkg/proxy/integrations/httpparser/httpparser.go

```diff
@@ -110,122 +110,183 @@ func ProcessOutgoingHttp(request []byte, clientConn, destConn net.Conn, h *hooks
-			return
+
+		// destConn is nil in case of test mode.
+		if destConn != nil {
+			_, err = destConn.Write(requestChunked)
+			if err != nil {
+				logger.Error("failed to write request message to the destination server")
+				return err
+			}
 		}
 	}
+	return nil
 }
 
 // Handled chunked requests when transfer-encoding is given.
-func chunkedRequest(finalReq *[]byte, clientConn, destConn net.Conn, logger *zap.Logger, transferEncodingHeader string) {
+func chunkedRequest(finalReq *[]byte, clientConn, destConn net.Conn, logger *zap.Logger, transferEncodingHeader string) error {
 	if transferEncodingHeader == "chunked" {
 		for {
+			//TODO: we have to implement a way to read the buffer chunk wise according to the chunk size (chunk size comes in hexadecimal)
+			// because it can happen that some chunks come after 5 seconds.
 			clientConn.SetReadDeadline(time.Now().Add(5 * time.Second))
 			requestChunked, err := util.ReadBytes(clientConn)
-			if err != nil && err != io.EOF {
+			if err != nil {
 				if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
 					break
 				} else {
-					logger.Error("failed to read the response message from the destination server", zap.Error(err))
-					return
+					logger.Error("failed to read the response message from the destination server")
+					return err
 				}
 			}
+
 			*finalReq = append(*finalReq, requestChunked...)
-			_, err = destConn.Write(requestChunked)
-			if err != nil {
-				logger.Error("failed to write request message to the destination server", zap.Error(err))
-				return
+			// destConn is nil in case of test mode.
+			if destConn != nil {
+				_, err = destConn.Write(requestChunked)
+				if err != nil {
+					logger.Error("failed to write request message to the destination server")
+					return err
+				}
 			}
 
 			//check if the intial request is completed
 			if strings.HasSuffix(string(requestChunked), "0\r\n\r\n") {
-				break
+				return nil
 			}
 		}
 	}
+	return nil
 }
 
 // Handled chunked responses when content-length is given.
-func contentLengthResponse(finalResp *[]byte, clientConn, destConn net.Conn, logger *zap.Logger, contentLength int) {
+func contentLengthResponse(finalResp *[]byte, clientConn, destConn net.Conn, logger *zap.Logger, contentLength int) error {
+	isEOF := false
 	for contentLength > 0 {
 		//Set deadline of 5 seconds
 		destConn.SetReadDeadline(time.Now().Add(5 * time.Second))
 		resp, err := util.ReadBytes(destConn)
 		if err != nil {
-			//Check if the connection closed.
 			if err == io.EOF {
-				logger.Error("connection closed by the destination server", zap.Error(err))
-				break
+				isEOF = true
+				logger.Debug("recieved EOF, connection closed by the destination server")
+				if len(resp) == 0 {
+					break
+				}
 			} else if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
 				logger.Info("Stopped getting data from the connection", zap.Error(err))
```

Comment:
Shouldn't this be a debug log?

Reply:
I guess we should tell the user that may be due to slow internet, the response couldn't be completed in time, hence this error.

Reply:
Will user understand that this is because of slow internet through that log?

Reply:
Since it is a timeout error, it will also be reflected in its application. So I guess the user will be able to know

Reply:
Got it.